File: Difficulty: Medium/K Sized Subarray Maximum/k-sized-subarray-maximum.cpp
class Solution {
  public:
    vector<int> maxOfSubarrays(vector<int>& arr, int k) {
        // code here
       vector<int> ans;
      int maxv=INT_MIN, maxi=-1;
      
      int i=0, j=0, n=arr.size();
      
      for(j=0; j<n; j++){
        if(arr[j] > maxv) { maxv=arr[j]; maxi=j; }
        
        if((j-i+1) == k){
          ans.push_back(maxv);
          if(i==maxi){
            maxv=INT_MIN;
            i++;
            for(int m=i; m<=j; m++){
              if(arr[m] > maxv) { maxv=arr[m]; maxi=m; }    
            }
          }
          else i++;
        }
      }
      
      return ans;
    }
};

File: Difficulty: Medium/K Sized Subarray Maximum/k-sized-subarray-maximum.java
class Solution {
    static ArrayList<Integer> maxOfSubarrays(int[] arr, int k) {
        // code here
        Deque<Integer>dq=new LinkedList<>();
        ArrayList<Integer>ans=new ArrayList<>();
        int n=arr.length;
        for(int i=0;i<n;i++){
            if(!dq.isEmpty() && dq.peekFirst()==i-k){
                dq.pollFirst();
            }
            while(!dq.isEmpty() && arr[dq.peekLast()]<=arr[i]){
                dq.pollLast();
            }
            dq.offerLast(i);
            if(i>=k-1){
                ans.add(arr[dq.peekFirst()]);
            }
        }
        return ans;
    }
}